File: src/memos/memories/textual/hierarchical_markdown/summarizer.py
"""Depth-aware LLM summarisation with three-level escalation.

The summarizer reads children ``.md`` files, feeds their content to an LLM
with a depth-appropriate prompt, and returns the summary text + metadata.
If the summary is too long, it escalates to a more aggressive prompt, then
falls back to deterministic truncation.
"""

import json

from typing import Any

from memos.log import get_logger
from memos.memories.textual.hierarchical_markdown.fs import read_md
from memos.templates.hierarchical_markdown_prompts import (
    AGGRESSIVE_SUMMARY_PROMPT,
    CONDENSED_D1_PROMPT,
    CONDENSED_D2_PLUS_PROMPT,
    LEAF_SUMMARY_PROMPT,
    UPDATE_SUMMARY_PROMPT,
)


logger = get_logger(__name__)


def _parse_json_response(text: str) -> dict[str, Any]:
    """Best-effort JSON extraction from LLM output."""
    # Find the first { and last }
    start = text.find("{")
    end = text.rfind("}")
    if start < 0 or end < 0 or end <= start:
        return {}
    fragment = text[start : end + 1]
    # Strip markdown code fences
    fragment = fragment.replace("```json", "").replace("```", "")
    try:
        return json.loads(fragment)
    except json.JSONDecodeError:
        return {}


class HierarchicalSummarizer:
    """Produce condensed ``_summary.md`` content from children nodes."""

    def __init__(
        self,
        llm: Any,
        embedder: Any,
        condensed_target_tokens: int = 512,
    ) -> None:
        self.llm = llm
        self.embedder = embedder
        self.condensed_target_tokens = condensed_target_tokens

    def summarize(
        self,
        children_paths: list[str],
        depth: int,
        previous_summary: str | None = None,
    ) -> tuple[str, str, list[str]]:
        """Summarise children into a condensed node.

        Args:
            children_paths: Absolute paths to children ``.md`` files.
            depth: The depth of the new condensed node (children are depth-1).
            previous_summary: If re-summarising, the old summary text.

        Returns:
            ``(summary_body, key, tags)``
        """
        children_text = self._read_children(children_paths)

        # Pick prompt by depth
        if previous_summary is not None:
            prompt_template = UPDATE_SUMMARY_PROMPT
        elif depth == 1:
            prompt_template = LEAF_SUMMARY_PROMPT
        elif depth == 2:
            prompt_template = CONDENSED_D1_PROMPT
        else:
            prompt_template = CONDENSED_D2_PLUS_PROMPT

        # Level 1: Normal summarisation
        result = self._run_llm(prompt_template, children_text, previous_summary, temperature=0.2)
        if result and self._is_acceptable(result["summary"]):
            return result["summary"], result["key"], result.get("tags", [])

        # Level 2: Aggressive prompt
        logger.info("Escalating to aggressive summary (depth=%d)", depth)
        result = self._run_llm(
            AGGRESSIVE_SUMMARY_PROMPT, children_text, previous_summary, temperature=0.1
        )
        if result and self._is_acceptable(result["summary"]):
            return result["summary"], result["key"], result.get("tags", [])

        # Level 3: Deterministic truncation fallback
        logger.warning("Falling back to truncation (depth=%d)", depth)
        return self._truncate(children_text), "summary", []

    def summarize_for_key(self, text: str) -> tuple[str, list[str]]:
        """Lightweight: extract just a key and tags from text (no full summary)."""
        prompt = (
            "Given this text, produce a short descriptive key (5-15 words) and 1-5 tags.\n\n"
            f"<text>\n{text[:2000]}\n</text>\n\n"
            'Return ONLY valid JSON: {"key": "...", "tags": ["..."]}'
        )
        try:
            response = self.llm.generate([{"role": "user", "content": prompt}])
            data = _parse_json_response(response)
            return data.get("key", text[:60]), data.get("tags", [])
        except Exception:
            logger.warning("Key extraction failed, using fallback")
            return text[:60].replace("\n", " "), []

    def _run_llm(
        self,
        template: str,
        children_text: str,
        previous_summary: str | None,
        temperature: float,
    ) -> dict[str, Any] | None:
        """Fill prompt template and call LLM."""
        previous_context = ""
        if previous_summary:
            previous_context = (
                f"The previous summary was:\n<previous>\n{previous_summary}\n</previous>\n"
                "Update it based on the new children content."
            )

        prompt = template.replace("${conversation}", children_text)
        prompt = prompt.replace("${children_summaries}", children_text)
        prompt = prompt.replace("${content}", children_text)
        prompt = prompt.replace("${previous_summary}", previous_summary or "")
        prompt = prompt.replace("${previous_context}", previous_context)
        prompt = prompt.replace("${target_tokens}", str(self.condensed_target_tokens))

        try:
            response = self.llm.generate([{"role": "user", "content": prompt}])
            data = _parse_json_response(response)
            if "summary" in data and "key" in data:
                return data
            # Sometimes LLM puts summary in "content" or similar
            if "key" in data:
                data.setdefault("summary", data.get("content", children_text[:500]))
                return data
        except Exception:
            logger.exception("LLM summarisation failed")

        return None

    def _read_children(self, paths: list[str]) -> str:
        """Concatenate children content for the prompt."""
        parts: list[str] = []
        for path in paths:
            try:
                meta, body, _edges = read_md(path)
                key = meta.get("key", "")
                header = f"[{key}]" if key else ""
                parts.append(f"{header}\n{body}")
            except Exception:
                logger.warning("Could not read child: %s", path)
        return "\n\n---\n\n".join(parts)

    def _is_acceptable(self, summary: str) -> bool:
        """Check if the summary is within target token bounds."""
        estimated = len(summary) // 4
        # Allow 50% overshoot before escalating
        return estimated <= self.condensed_target_tokens * 1.5

    def _truncate(self, text: str) -> str:
        """Deterministic truncation fallback."""
        max_chars = self.condensed_target_tokens * 4
        if len(text) <= max_chars:
            return text
        return text[:max_chars].rsplit(" ", 1)[0] + "..."

File: apps/ayanamimem-openclaw/src/summarizer.ts
/** LLM-based summarisation via OpenAI-compatible API.
 *
 * Port of Python src/.../hierarchical_markdown/summarizer.py
 * All prompts are TS string constants matching hierarchical_markdown_prompts.py.
 */

import * as fs from "node:fs";
import { readMd } from "./fs-store.js";
import type { Logger } from "./types.js";

// ── Prompt templates ──────────────────────────────────────────────────────────

const LEAF_KEY_PROMPT = `Given this conversation chunk, produce a short descriptive key (5-15 words) and 1-5 tags.

<conversation>
{{conversation}}
</conversation>

Return ONLY valid JSON: {"key": "...", "tags": ["..."]}`;

const LEAF_SUMMARY_PROMPT = `You are a memory summariser. Below is a raw conversation chunk.

<conversation>
{{conversation}}
</conversation>

{{previous_context}}

Produce a concise narrative summary (max {{target_tokens}} tokens) capturing key facts, decisions, preferences, and action items. Also produce a short key (5-15 words) and 1-5 tags.

Return ONLY valid JSON: {"key": "...", "summary": "...", "tags": ["..."]}`;

const CONDENSED_D1_PROMPT = `You are a memory summariser. Below are summaries of related conversation segments.

<summaries>
{{children_summaries}}
</summaries>

{{previous_context}}

Produce a thematic summary (max {{target_tokens}} tokens) preserving key facts, names, numbers, and decisions. Also produce a key (5-15 words) and 1-5 tags.

Return ONLY valid JSON: {"key": "...", "summary": "...", "tags": ["..."]}`;

const CONDENSED_D2_PLUS_PROMPT = `You are a memory summariser working at high abstraction. Below are summaries of broad topic groups.

<summaries>
{{children_summaries}}
</summaries>

{{previous_context}}

Produce an abstract overview (max {{target_tokens}} tokens) identifying major themes and patterns. Also produce a key (5-15 words) and 1-5 tags.

Return ONLY valid JSON: {"key": "...", "summary": "...", "tags": ["..."]}`;

const AGGRESSIVE_SUMMARY_PROMPT = `You are a memory summariser under strict space constraints. Compress the following.

<content>
{{content}}
</content>

Keep ONLY concrete facts (names, numbers, dates), firm decisions, preferences, and action items. Max {{target_tokens}} tokens.

Return ONLY valid JSON: {"key": "...", "summary": "...", "tags": ["..."]}`;

// ── JSON extraction ───────────────────────────────────────────────────────────

function parseJson(text: string): Record<string, unknown> {
  const start = text.indexOf("{");
  const end   = text.lastIndexOf("}");
  if (start < 0 || end < 0) return {};
  try {
    return JSON.parse(text.slice(start, end + 1));
  } catch {
    return {};
  }
}

// ── Summarizer class ──────────────────────────────────────────────────────────

interface SummarizerCfg {
  chatApiBase: string;
  chatApiKey: string;
  chatModel: string;
  targetTokens?: number;
}

export class Summarizer {
  private cfg: SummarizerCfg;
  private log: Logger;
  private targetTokens: number;

  constructor(cfg: SummarizerCfg, log: Logger) {
    this.cfg = cfg;
    this.log = log;
    this.targetTokens = cfg.targetTokens ?? 512;
  }

  /** Lightweight: extract just a key and tags from raw text (no full summary). */
  async extractKey(text: string): Promise<{ key: string; tags: string[] }> {
    const prompt = LEAF_KEY_PROMPT.replace("{{conversation}}", text.slice(0, 2000));
    try {
      const resp = await this._call(prompt);
      const data = parseJson(resp);
      return {
        key:  String(data.key  ?? text.slice(0, 60).replace(/\n/g, " ")),
        tags: Array.isArray(data.tags) ? data.tags.map(String) : [],
      };
    } catch {
      return { key: text.slice(0, 60).replace(/\n/g, " "), tags: [] };
    }
  }

  /** Summarise children .md files into a condensed node. */
  async summarize(
    childPaths: string[],
    depth: number,
    prevSummary?: string,
  ): Promise<{ summary: string; key: string; tags: string[] }> {
    const childrenText = this._readChildren(childPaths);

    const prevCtx = prevSummary
      ? `The previous summary was:\n<previous>\n${prevSummary}\n</previous>\nUpdate it based on the new children.`
      : "";

    const template = prevSummary
      ? CONDENSED_D1_PROMPT
      : depth === 1 ? LEAF_SUMMARY_PROMPT
      : depth === 2 ? CONDENSED_D1_PROMPT
      : CONDENSED_D2_PLUS_PROMPT;

    // Level 1: normal summarisation
    let result = await this._runPrompt(template, childrenText, prevCtx, 0.2);
    if (result && this._isAcceptable(result.summary)) return result;

    // Level 2: aggressive
    this.log.warn("Escalating to aggressive summary (depth=%d)", depth);
    result = await this._runPrompt(AGGRESSIVE_SUMMARY_PROMPT, childrenText, prevCtx, 0.1);
    if (result && this._isAcceptable(result.summary)) return result;

    // Level 3: truncation fallback
    this.log.warn("Using truncation fallback (depth=%d)", depth);
    const truncated = childrenText.slice(0, this.targetTokens * 4);
    return { summary: truncated, key: "summary", tags: [] };
  }

  // ── Private helpers ─────────────────────────────────────────────────────────

  private async _runPrompt(
    template: string,
    childrenText: string,
    prevCtx: string,
    _temperature: number,
  ): Promise<{ summary: string; key: string; tags: string[] } | null> {
    const prompt = template
      .replace("{{conversation}}", childrenText)
      .replace("{{children_summaries}}", childrenText)
      .replace("{{content}}", childrenText)
      .replace("{{previous_summary}}", prevCtx)
      .replace("{{previous_context}}", prevCtx)
      .replace(/\{\{target_tokens\}\}/g, String(this.targetTokens));

    try {
      const resp = await this._call(prompt);
      const data = parseJson(resp);
      if (data.summary && data.key) {
        return {
          summary: String(data.summary),
          key: String(data.key),
          tags: Array.isArray(data.tags) ? data.tags.map(String) : [],
        };
      }
    } catch (e) {
      this.log.warn("LLM summarisation failed: %s", String(e));
    }
    return null;
  }

  private _readChildren(paths: string[]): string {
    const parts: string[] = [];
    for (const p of paths) {
      if (!fs.existsSync(p)) continue;
      try {
        const { meta, body } = readMd(p);
        const header = meta.key ? `[${meta.key}]` : "";
        parts.push(`${header}\n${body}`);
      } catch {
        this.log.warn("Could not read child: %s", p);
      }
    }
    return parts.join("\n\n---\n\n");
  }

  private _isAcceptable(summary: string): boolean {
    return summary.length / 4 <= this.targetTokens * 1.5;
  }

  private async _call(prompt: string): Promise<string> {
    const apiBase = this.cfg.chatApiBase.replace(/\/$/, "");
    const res = await fetch(`${apiBase}/chat/completions`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${this.cfg.chatApiKey}`,
      },
      body: JSON.stringify({
        model: this.cfg.chatModel,
        messages: [{ role: "user", content: prompt }],
        temperature: 0.2,
      }),
      signal: AbortSignal.timeout(60_000),
    });
    if (!res.ok) throw new Error(`LLM API ${res.status}: ${await res.text()}`);
    const json = (await res.json()) as { choices: { message: { content: string } }[] };
    return json.choices[0]?.message?.content ?? "";
  }
}
